java/ql/src/experimental/Security/CWE/CWE-117/LogInjection.ql

```diff
@@ -0,0 +1,60 @@
+/**
+ * @name Log Injection
+ * @description Building log entries from user-controlled sources is vulnerable to
+ *              insertion of forged log entries by a malicious user.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
+ * @id java/log-injection
+ * @tags security
+ *       external/cwe/cwe-117
+ */
+
+import java
+import DataFlow::PathGraph
+import experimental.semmle.code.java.Logging
+import semmle.code.java.dataflow.FlowSources
+
+/**
+ * A taint-tracking configuration for tracking untrusted user input used in log entries.
+ */
+private class LogInjectionConfiguration extends TaintTracking::Configuration {
+  LogInjectionConfiguration() { this = "Log Injection" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(LoggingCall c | sink.asExpr() = c.getALogArgument())
+  }
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    node.getType() instanceof BoxedType or node.getType() instanceof PrimitiveType
```

Comment:
What about calls to `String replace()`?

Reply:
Sanitizing using `replace(...)` (or `replaceAll(...)`) can be error prone. E.g. only removing `\n` might still cause issues for other Unicode control characters such as `\r`, `\0` or `\a` (bell character).

Maybe it should actually be the task of the logging framework to sanitize log messages?

Reply:
@Marcono1234 ideally the framework should manage it but from what I can find, they don't necessarily sanitize the new line character. So, @owen-mc I think we should not include calls to `replace` here. Let's see the output of the LGTM run. If there are a lot of FP's due to this, I will add an appropriate sanitizer.